File: src/commonMain/kotlin/com/scoreo/application/GetHeadToHeadUseCase.kt
package com.scoreo.application

import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.Match
import com.scoreo.domain.port.GameTypeRepository
import com.scoreo.domain.port.MatchRepository
import com.scoreo.domain.port.PlayerRepository

class GetHeadToHeadUseCase(
    private val matchRepository: MatchRepository,
    private val gameTypeRepository: GameTypeRepository,
    private val playerRepository: PlayerRepository,
    private val eloCalculator: EloCalculator = EloCalculator(),
) {
    operator fun invoke(gameTypeId: String? = null): List<PlayerDetail> {
        val gameTypes = gameTypeRepository.getAll().associateBy { it.id }
        val players = playerRepository.getAll(includeInactive = true).associateBy { it.id }
        val allMatches = matchRepository.getAll()
            .let { matches -> if (gameTypeId != null) matches.filter { it.gameTypeId == gameTypeId } else matches }

        val totalWins = mutableMapOf<String, Int>()
        val totalLosses = mutableMapOf<String, Int>()
        val h2h = mutableMapOf<Pair<String, String>, MutableList<Int>>()

        var orphanedMatches = 0
        allMatches.forEach { match ->
            val gameType = gameTypes[match.gameTypeId]
            if (gameType == null) {
                orphanedMatches++
                return@forEach
            }
            val winners = match.getWinners(gameType).toSet()
            val participants = match.playerScores.map { it.playerId }

            participants.forEach { pid ->
                if (pid in winners) {
                    totalWins[pid] = (totalWins[pid] ?: 0) + 1
                } else {
                    totalLosses[pid] = (totalLosses[pid] ?: 0) + 1
                }
            }

            for (i in participants.indices) {
                for (j in i + 1 until participants.size) {
                    val a = participants[i]
                    val b = participants[j]
                    val key = if (a < b) Pair(a, b) else Pair(b, a)
                    val h = h2h.getOrPut(key) { mutableListOf(0, 0) }
                    if (a in winners && b !in winners) {
                        if (key.first == a) h[0] = h[0] + 1 else h[1] = h[1] + 1
                    }
                    if (b in winners && a !in winners) {
                        if (key.first == b) h[0] = h[0] + 1 else h[1] = h[1] + 1
                    }
                }
            }
        }

        if (orphanedMatches > 0) {
            println("[Scoreo] Warning: $orphanedMatches match(es) reference non-existent game types")
        }

        val eloMap = eloCalculator.compute(allMatches, gameTypes)

        return players.entries.mapNotNull { (pid, player) ->
            val wins = totalWins[pid] ?: 0
            val losses = totalLosses[pid] ?: 0
            if (wins + losses == 0) return@mapNotNull null

            val headToHead = h2h.filterKeys { (a, b) -> a == pid || b == pid }.map { (pair, wl) ->
                val isA = pair.first == pid
                val oppId = if (isA) pair.second else pair.first
                val oppName = players[oppId]?.name ?: oppId
                HeadToHeadEntry(
                    opponentId = oppId,
                    opponentName = oppName,
                    wins = if (isA) wl[0] else wl[1],
                    losses = if (isA) wl[1] else wl[0],
                )
            }.sortedByDescending { it.wins }

            PlayerDetail(
                playerId = pid,
                name = player.name,
                wins = wins,
                losses = losses,
                elo = eloMap[pid] ?: 1200,
                headToHead = headToHead,
            )
        }.sortedByDescending { it.elo }
    }
}

File: src/application/getHeadToHeadUseCase.ts
import { getWinners } from '../domain/model/match'
import type { GameTypeRepository } from '../domain/port/gameTypeRepository'
import type { MatchRepository } from '../domain/port/matchRepository'
import type { PlayerRepository } from '../domain/port/playerRepository'
import { EloCalculator } from './eloCalculator'

export interface HeadToHeadEntry {
  opponentId: string
  opponentName: string
  wins: number
  losses: number
}

export interface PlayerDetail {
  playerId: string
  name: string
  wins: number
  losses: number
  elo: number
  headToHead: HeadToHeadEntry[]
}

interface HeadToHeadTally {
  first: string
  second: string
  winsFirst: number
  winsSecond: number
}

export class GetHeadToHeadUseCase {
  constructor(
    private readonly matchRepository: MatchRepository,
    private readonly gameTypeRepository: GameTypeRepository,
    private readonly playerRepository: PlayerRepository,
    private readonly eloCalculator: EloCalculator = new EloCalculator(),
  ) {}

  invoke(gameTypeId?: string): PlayerDetail[] {
    const gameTypes = new Map(this.gameTypeRepository.getAll().map((gt) => [gt.id, gt]))
    const players = new Map(this.playerRepository.getAll(true).map((p) => [p.id, p]))
    const allMatches = this.matchRepository
      .getAll()
      .filter((m) => gameTypeId === undefined || m.gameTypeId === gameTypeId)

    const totalWins = new Map<string, number>()
    const totalLosses = new Map<string, number>()
    const h2h = new Map<string, HeadToHeadTally>()

    for (const match of allMatches) {
      const gameType = gameTypes.get(match.gameTypeId)
      if (!gameType) continue
      const winners = new Set(getWinners(match, gameType))
      const participants = match.playerScores.map((s) => s.playerId)

      for (const pid of participants) {
        if (winners.has(pid)) {
          totalWins.set(pid, (totalWins.get(pid) ?? 0) + 1)
        } else {
          totalLosses.set(pid, (totalLosses.get(pid) ?? 0) + 1)
        }
      }

      for (let i = 0; i < participants.length; i++) {
        for (let j = i + 1; j < participants.length; j++) {
          const a = participants[i]
          const b = participants[j]
          const [first, second] = a < b ? [a, b] : [b, a]
          const key = `${first} ${second}`
          const tally = h2h.get(key) ?? { first, second, winsFirst: 0, winsSecond: 0 }
          if (winners.has(a) && !winners.has(b)) {
            if (first === a) tally.winsFirst += 1
            else tally.winsSecond += 1
          }
          if (winners.has(b) && !winners.has(a)) {
            if (first === b) tally.winsFirst += 1
            else tally.winsSecond += 1
          }
          h2h.set(key, tally)
        }
      }
    }

    const eloMap = this.eloCalculator.compute(allMatches, gameTypes)

    const details: PlayerDetail[] = []
    for (const [pid, player] of players) {
      const wins = totalWins.get(pid) ?? 0
      const losses = totalLosses.get(pid) ?? 0
      if (wins + losses === 0) continue

      const headToHead: HeadToHeadEntry[] = []
      for (const { first, second, winsFirst, winsSecond } of h2h.values()) {
        if (first !== pid && second !== pid) continue
        const isFirst = first === pid
        const oppId = isFirst ? second : first
        const oppName = players.get(oppId)?.name ?? oppId
        headToHead.push({
          opponentId: oppId,
          opponentName: oppName,
          wins: isFirst ? winsFirst : winsSecond,
          losses: isFirst ? winsSecond : winsFirst,
        })
      }
      headToHead.sort((a, b) => b.wins - a.wins)

      details.push({
        playerId: pid,
        name: player.name,
        wins,
        losses,
        elo: eloMap.get(pid) ?? 1200,
        headToHead,
      })
    }

    return details.sort((a, b) => b.elo - a.elo)
  }
}
